### PR TITLE
agent label assignment and matching

### PR DESCRIPTION
**Pull request by logikone**
* allow matching against stages with specific tags
* agent/planner labels are set via the `DRONE_AGENT_LABELS` environment
variable

### CODE REVIEW

**Comment by bradrydzewski** [comment]
Thanks Chris, this looks really good. I should have some time tomorrow morning to review in greater detail and get it merged.

**Comment by logikone** [comment]
right on, ty sir!

**Comment by logikone** [comment]
i noticed a typo in my commit message so i amended it. `s/LABLES/LABELS`

File: config/config.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ type (
 			Kernel      string
 			Environ     []string
 			Volumes     []string
+			Labels      []string `envconfig:"DRONE_AGENT_LABELS"`
```
Comment (msaizar) [inline config/config.go]: I've been working on something similar, used 
```go
Labels      map[string]string `envconfig:"DRONE_AGENT_LABELS"`
```

instead, `envconfig` will handle `DRONE_AGENT_LABELS=mem:high,cpu:load`.

File: engine/install.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ type installer struct {
 	keepaliveTime    time.Duration
 	keepaliveTimeout time.Duration
 	runner           config.Runner
+	labels 			 []string
```
Comment (msaizar) [inline engine/install.go]: Same here with `map[string]string`

File: engine/planner.go [hunk 3]
```diff
@@ -29,6 +30,7 @@ type planner struct {
 	max     int           // max number of servers to allocate
 	cap     int           // capacity per-server
 	ttu     time.Duration // minimum server age
+	labels  []string
```
Comment (msaizar) [inline engine/planner.go]: And here

File: engine/planner.go [hunk 4]
```diff
@@ -268,8 +270,29 @@ func (p *planner) listBusy(ctx context.Context) (map[string]struct{}, error) {
...
+
 	return stage.OS == p.os &&
 		stage.Arch == p.arch &&
 		stage.Variant == p.version &&
-		stage.Kernel == p.kernel
+		stage.Kernel == p.kernel &&
+		labelMatch
+}
+
+func (p *planner) matchLabel(stage *drone.Stage) bool {
```
Comment (msaizar) [inline engine/planner.go]: This wouldn't be necessary

File: engine/planner.go [hunk 5]
```diff
@@ -268,8 +270,29 @@ func (p *planner) listBusy(ctx context.Context) (map[string]struct{}, error) {
...
+	if len(p.labels) > 0 || len(stage.Labels) > 0 {
+		labelMatch = p.matchLabel(stage)
+	}
+
 	return stage.OS == p.os &&
 		stage.Arch == p.arch &&
 		stage.Variant == p.version &&
-		stage.Kernel == p.kernel
+		stage.Kernel == p.kernel &&
+		labelMatch
```
Comment (msaizar) [inline engine/planner.go]: We could match with `reflect.DeepEqual(stage.Labels, p.labels)` instead

File: engine/install.go [hunk 6]
```diff
@@ -158,6 +159,12 @@ poller:
 		fmt.Sprintf("DRONE_RUNNER_PRIVILEGED_IMAGES=%s", i.runner.Privileged),
 	)
 
+	if len(i.labels) > 0 {
+		envs = append(envs,
+			fmt.Sprintf("DRONE_RUNNER_LABELS=%s", strings.Join(i.labels, ",")),
```
Comment (msaizar) [inline engine/install.go]: Not 100% sure here, I think with the `map` we could just do `fmt.Sprintf("DRONE_RUNNER_LABELS=%s", i.labels),` in the big `append` above

File: config/config.go (hunk 1 above)
Comment (bradrydzewski) [reply config/config.go]: this is true.  we did not use maps for environment variables (not sure why) but we did use maps for other configuration parameters. I think in this case using a map makes sense.

File: engine/planner.go [hunk 7]
```diff
@@ -268,8 +270,29 @@ func (p *planner) listBusy(ctx context.Context) (map[string]struct{}, error) {
...
+}
+
+func (p *planner) matchLabel(stage *drone.Stage) bool {
+	plannerLabels := formatLabels(p.labels)
+
+	for key, val := range stage.Labels {
+		if plannerVal, ok := plannerLabels[key]; ok {
+			if strings.EqualFold(plannerVal, val) {
+				return true
```
Comment (bradrydzewski) [inline engine/planner.go]: this should return false if a single label does not match, and should return true only if all labels match.

File: engine/planner.go (hunk 5 above)
Comment (bradrydzewski) [reply engine/planner.go]: this probably comes down to personal preference, but I prefer the current approach of using a typed function.

File: engine/install.go (hunk 6 above)
Comment (bradrydzewski) [reply engine/install.go]: if you have `map[string]string{"foo":"bar"}` it would print `map[foo:bar]`

Comment (msaizar) [reply engine/install.go]: Thanks, I hadn't tested this bit, completely new to `go`. Sorry for the bad advice!

File: config/config.go (hunk 1 above)
Comment (logikone) [reply config/config.go]: nice, i didn't realize `envconfig` handled maps. i'll get these all updated

File: engine/planner.go (hunk 7 above)
Comment (logikone) [reply engine/planner.go]: ok cool, ya i wasn't sure on that one. i'll use `reflect.DeepEqual` like @msaizar mentioned above

Comment (bradrydzewski) [reply engine/planner.go]: I recommend copying the exact same function we are using to match on the server, to ensure the logic syncs up

https://github.com/drone/drone/blob/master/scheduler/queue/queue.go#L232
https://github.com/drone/drone/blob/master/scheduler/queue/queue.go#L175

Comment (logikone) [reply engine/planner.go]: ah, ok. will do

Comment (logikone) [reply engine/planner.go]: ok, i've got that updated

File: engine/formatlabels.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package engine
+
+import (
+	"strings"
+)
+
+// formatLabels takes a comma separated list of agent labels
+// and formats it into map[string]string for agent label matching
+func formatLabels(labels []string) map[string]string {
```
Comment (bradrydzewski) [inline engine/formatlabels.go]: is this still being used?

**Comment by bradrydzewski** [comment]
thanks Chris for the changes. Sorry to pull you onto your computer on a Saturday!

File: engine/formatlabels.go (hunk 8 above)
Comment (logikone) [reply engine/formatlabels.go]: no, i was just realizing that myself. i'll pull this out right now

**Comment by logikone** [comment]
no worries, thank you for the input!

**Pull request merged**